### PR TITLE
Feature/add html converter service

### PR DESCRIPTION
**Pull request by thvrijn**
## Description

Describe your changes in detail.

## Motivation and context

Why is this change required? What problem does it solve?

If it fixes an open issue, please link to the issue here (if you write `fixes #num`
or `closes #num`, the issue will be automatically closed when the pull is accepted.)

## How has this been tested?

Please describe in detail how you tested your changes.

Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc.

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Go over all the following points, and put an `x` in all the boxes that apply.

Please, please, please, don't send your pull request until all of the boxes are ticked. Once your pull request is created, it will trigger a build on our [continuous integration](http://www.phptherightway.com/#continuous-integration) server to make sure your [tests and code style pass](https://help.github.com/articles/about-required-status-checks/).

- [ ] I have read the **[CONTRIBUTING](CONTRIBUTING.md)** document.
- [ ] My pull request addresses exactly one patch/feature.
- [ ] I have created a branch for this patch/feature.
- [ ] Each individual commit in the pull request is meaningful.
- [ ] I have added tests to cover my changes.
- [ ] If my change requires a change to the documentation, I have updated it accordingly.

If you're unsure about any of these, don't hesitate to ask. We're here to help!

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -25,6 +25,9 @@
     "require": {
         "php": "^7.1.3",
         "ext-json": "*",
+        "altelma/laravel-signrequest": "dev-master",
```
Comment (tychovbh) [inline composer.json]: deze mag weg via compose remove

File: config/mvc-html-converter.php [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+return [
+    'default' => env('HTML_CONVERTER_SERVICE', 'PhantomMagickConverter'),
+    'providers' => [
```
Comment (tychovbh) [inline config/mvc-html-converter.php]: als je provider nog geen config verwacht zoals tokens of urls etc. dan mag dit weg.

File: config/mvc-address-lookup.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+return [
+    'default' => env('ADDRESS_LOOKUP_SERICE', 'PdokService'),
+    'providers' => [
```
Comment (tychovbh) [inline config/mvc-address-lookup.php:5]: als je provider nog geen config verwacht zoals tokens of urls etc. dan mag dit weg.

Ik bedenk me nu wel dat we hier requests doen naar een url, dus ik zou die base url hiernaar toe verplaatsen. Als pdok dan een x ze url veranderd kunnen we dat via de config aanpassen inplaats van in de code.

File: config/mvc-document-sign.php [hunk 4]
```diff
@@ -1,9 +1,9 @@
 <?php
 
 return [
-    'default' => 'signrequest',
+    'default' => env('DOCUMENT_SIGN_PROVIDER', 'SignRequest'),
     'providers' => [
-        'signrequest' => [
+        'SignRequest' => [
```
Comment (tychovbh) [inline config/mvc-document-sign.php:6]: token en subdomein moeten uit .env komen.

File: src/Contract.php [hunk 5]
```diff
@@ -22,8 +29,35 @@ class Contract extends Model
...
     public function __construct(array $attributes = [])
     {
-        $this->fillables('file', 'status', 'signed_at', 'options');
-        $this->columns('file', 'status', 'signed_at', 'options');
+        $this->fillables('file', 'status', 'signed_at', 'options', 'template', 'external_id', 'user_id');
+        $this->columns('file', 'status', 'signed_at', 'options', 'external_id', 'user_id');
         parent::__construct($attributes);
     }
+
+    public function toPdf(HtmlConverterInterface $htmlConverter)
```
Comment (tychovbh) [inline src/Contract.php:45]: missing doc

File: src/Contract.php [hunk 6]
```diff
@@ -2,6 +2,13 @@
 
 namespace Tychovbh\Mvc;
 
+use GuzzleHttp\Client;
+use Illuminate\Support\Str;
+use Tychovbh\Mvc\Services\DocumentSign\DocumentSignInterface;
+use Tychovbh\Mvc\Services\DocumentSign\SignRequest;
+use Tychovbh\Mvc\Services\HtmlConverter\HtmlConverterInterface;
+use Tychovbh\Mvc\Services\HtmlConverter\PhantomMagickConverter;
```
Comment (tychovbh) [inline src/Contract.php]: deze wordt niet gebruikt?

File: src/Contract.php [hunk 7]
```diff
@@ -22,8 +29,35 @@ class Contract extends Model
      */
     public function __construct(array $attributes = [])
     {
-        $this->fillables('file', 'status', 'signed_at', 'options');
-        $this->columns('file', 'status', 'signed_at', 'options');
+        $this->fillables('file', 'status', 'signed_at', 'options', 'template', 'external_id', 'user_id');
+        $this->columns('file', 'status', 'signed_at', 'options', 'external_id', 'user_id');
```
Comment (tychovbh) [inline src/Contract.php]: id mag hier bij

File: src/Contract.php [hunk 8]
```diff
@@ -22,8 +29,35 @@ class Contract extends Model
...
+
+        $page = view($this->template);
+        $html = $page->render();
+        $path = 'contracts/contract.pdf';
+        $htmlConverter->page($html)->save($path);
+        $this->file = $path;
+        $this->unsetAttribute('template');
+    }
+
+    public function sign(DocumentSignInterface $documentSign)
```
Comment (tychovbh) [inline src/Contract.php:63]: missing doc

File: src/Contract.php [hunk 9]
```diff
@@ -22,8 +29,35 @@ class Contract extends Model
...
+    {
+        if (!$this->file) {
```
Comment (tychovbh) [inline src/Contract.php]: add a config to mvc-contracts: document-sign.enabled: true/false (filled via .env)
check here in the if statement for the enabling of the service

File: src/Contract.php [hunk 10]
```diff
@@ -22,8 +29,35 @@ class Contract extends Model
...
-        $this->fillables('file', 'status', 'signed_at', 'options');
-        $this->columns('file', 'status', 'signed_at', 'options');
+        $this->fillables('file', 'status', 'signed_at', 'options', 'template', 'external_id', 'user_id');
+        $this->columns('file', 'status', 'signed_at', 'options', 'external_id', 'user_id');
         parent::__construct($attributes);
     }
+
+    public function toPdf(HtmlConverterInterface $htmlConverter)
+    {
+        if (!$this->template) {
```
Comment (tychovbh) [inline src/Contract.php]: add a config to mvc-contracts: pdf.enabled: true/false (filled via .env)
check here in the if statement for the enabling of the service

File: src/Http/Controllers/AddressController.php [hunk 11]
```diff
@@ -18,7 +18,7 @@ public function store(Request $request): JsonResource
         try {
             return new $this->resource($this->repository::withParams($request->toArray())->first());
         } catch (ModelNotFoundException $exception) {
-            return parent::store($request);
+             return parent::store($request);
```
Comment (tychovbh) [inline src/Http/Controllers/AddressController.php]: looks like this is a mistake?

File: src/Model.php [hunk 12]
```diff
@@ -351,4 +351,13 @@ public function hasColumn(string $key): bool
 
         return in_array($key, $this->columns);
     }
+
+    /**
+     * Unsets an attribute
+     * @param $key
+     */
+    public function unsetAttribute($key)
```
Comment (tychovbh) [inline src/Model.php]: $key should be of type "string"

File: src/MvcServiceProvider.php [hunk 13]
```diff
@@ -21,8 +21,13 @@
 use Tychovbh\Mvc\Http\Middleware\ValidateMiddleware;
 use Tychovbh\Mvc\Http\Middleware\CacheMiddleware;
 use Tychovbh\Mvc\Observers\AddressObserver;
+use Tychovbh\Mvc\Observers\ContractObserver;
 use Tychovbh\Mvc\Observers\PaymentObserver;
+use Tychovbh\Mvc\Services\AddressLookup\AddressLookupInterface;
+use Tychovbh\Mvc\Services\DocumentSign\DocumentSignInterface;
+use Tychovbh\Mvc\Services\DocumentSign\DocuSign;
```
Comment (tychovbh) [inline src/MvcServiceProvider.php]: unused class?

File: src/MvcServiceProvider.php [hunk 14]
```diff
@@ -21,8 +21,13 @@
...
 use Tychovbh\Mvc\Services\DocumentSign\SignRequest;
```
Comment (tychovbh) [inline src/MvcServiceProvider.php]: unused class?

File: src/Observers/AddressObserver.php [hunk 15]
```diff
@@ -4,18 +4,28 @@
...
+use Tychovbh\Mvc\Services\AddressLookup\AddressLookupInterface;
 use Tychovbh\Mvc\Services\AddressLookup\PdokService;
 
+/**
+ * @property AddressLookupInterface addressLookup
+ */
 class AddressObserver
 {
+
+    public function __construct(AddressLookupInterface $addressLookup)
```
Comment (tychovbh) [inline src/Observers/AddressObserver.php:19]: missing doc

File: src/Observers/ContractObserver.php [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+use Tychovbh\Mvc\Services\DocumentSign\DocumentSignInterface;
+use Tychovbh\Mvc\Services\HtmlConverter\HtmlConverterInterface;
+
+/**
+ * @property DocumentSignInterface documentSign
+ * @property HtmlConverterInterface htmlConverter
+ */
+class ContractObserver
+{
+    public function __construct(DocumentSignInterface $documentSign, HtmlConverterInterface $htmlConverter)
```
Comment (tychovbh) [inline src/Observers/ContractObserver.php:15]: missing doc all methods check

File: src/Services/DocumentSign/DocuSign.php [hunk 17]
```diff
@@ -0,0 +1,49 @@
+<?php
+namespace Tychovbh\Mvc\Services\DocumentSign;
+
+use GuzzleHttp\Client;
+use Illuminate\Http\UploadedFile;
+
+class DocuSign implements DocumentSignInterface
+{
+    public function signer(string $email, string $firstname = '', string $lastname = ''): DocumentSignInterface
```
Comment (tychovbh) [inline src/Services/DocumentSign/DocuSign.php:16]: Missing doc blocks whole file

File: src/Services/DocumentSign/DocumentSignInterface.php [hunk 18]
```diff
@@ -15,13 +15,21 @@ interface DocumentSignInterface
      */
     public function signer(string $email, string $firstname = '', string $lastname = ''): DocumentSignInterface;
 
+    /**
+     * Creates a Document
+     * @param string $path
+     * @param string|null $name
```
Comment (tychovbh) [inline src/Services/DocumentSign/DocumentSignInterface.php]: this is not null anymore

File: src/Services/DocumentSign/SignRequest.php [hunk 19]
```diff
@@ -50,18 +49,12 @@ public function signer(string $email, string $firstname = '', string $lastname =
...
     }
 
-    /**
-     * Creates a Document
-     * @param UploadedFile $file
-     * @param string|null $webhook
-     * @return array
-     */
-    public function create(UploadedFile $file, string $webhook = null): array
+    private function createRequest(string $file, $name, $webhook = null)
```
Comment (tychovbh) [inline src/Services/DocumentSign/SignRequest.php]: missing doc block, and name/webhook should be typehinted

File: src/Services/DocumentSign/SignRequest.php [hunk 20]
```diff
@@ -70,13 +63,35 @@ public function create(UploadedFile $file, string $webhook = null): array
                 'status' => Arr::get($response, 'status', '')
             ];
         } catch (\Exception $exception) {
-            error('SignRequestService create error', [
+            error('SignRequestService createFromUpload error', [
```
Comment (tychovbh) [inline src/Services/DocumentSign/SignRequest.php]: keep it at create

File: src/Services/DocumentSign/SignRequest.php [hunk 21]
```diff
@@ -70,13 +63,35 @@ public function create(UploadedFile $file, string $webhook = null): array
...
                 'message' => $exception->getMessage(),
                 'line' => $exception->getLine(),
             ]);
         }
     }
 
+    /**
+     * Creates a Document
+     * @param string $path
+     * @param string|null $name
```
Comment (tychovbh) [inline src/Services/DocumentSign/SignRequest.php]: remove null

File: src/Services/DocumentSign/SignRequest.php [hunk 22]
```diff
@@ -95,7 +110,7 @@ public function sign(string $id, string $from_name, string $from_email, string $
             $response = $this->request('post', '/signrequests', [
                 'from_email' => $from_email,
                 'from_email_name' => $from_name,
-                'document' => config('mvc-document-sign.providers.signrequest.subdomain') . '/documents/' . $id . '/',
+                'document' => config('mvc-document-sign.providers.SignRequest.subdomain') . '/documents/' . $id . '/',
```
Comment (tychovbh) [inline src/Services/DocumentSign/SignRequest.php]: i dont like the look of this change to:

- add class property private $config
- in constructor set property with config mvc-document.sign.providers.SignRequest
- here in the code use Arr::get to get the subdomain config. Doe this for all configs in this file

File: src/Services/HtmlConverter/HtmlConverterInterface.php [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+
+interface HtmlConverterInterface
+{
+    /**
+     * @param string $html
+     * @return mixed
+     */
+    public function page(string $html);
+
+    /**
```
Comment (tychovbh) [inline src/Services/HtmlConverter/HtmlConverterInterface.php:14]: missing description

File: src/Services/HtmlConverter/HtmlConverterInterface.php [hunk 24]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace Tychovbh\Mvc\Services\HtmlConverter;
+
+interface HtmlConverterInterface
+{
+    /**
```
Comment (tychovbh) [inline src/Services/HtmlConverter/HtmlConverterInterface.php:7]: missing descriptio

File: src/Services/HtmlConverter/HtmlConverterInterface.php [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+     * @param string $html
+     * @return mixed
```
Comment (tychovbh) [inline src/Services/HtmlConverter/HtmlConverterInterface.php]: this should return $this, so $return HtmlConverterInterface
als add this to the method page(): HtmlConverterInterface;

File: src/Services/HtmlConverter/HtmlConverterInterface.php [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+     */
+    public function page(string $html);
+
+    /**
+     * @param string $path
+     * @param string $type
+     * @return mixed
```
Comment (tychovbh) [inline src/Services/HtmlConverter/HtmlConverterInterface.php]: i would like this to return boolean true/false

File: src/Services/HtmlConverter/PhantomMagickConverter.php [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+use Anam\PhantomMagick\Converter;
+
+class PhantomMagickConverter implements HtmlConverterInterface
+{
+    /**
+     * @var Collection
+     */
+    public $pages;
+
+    public function __construct()
```
Comment (tychovbh) [inline src/Services/HtmlConverter/PhantomMagickConverter.php:18]: missing doc

File: src/Services/HtmlConverter/PhantomMagickConverter.php [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        $this->pages = collect([]);
+    }
+
+    /**
+     * @param string $html
+     * @return Collection|mixed
```
Comment (tychovbh) [inline src/Services/HtmlConverter/PhantomMagickConverter.php]: returns this not a collection

File: src/Services/HtmlConverter/PhantomMagickConverter.php [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    public function page(string $html)
```
Comment (tychovbh) [inline src/Services/HtmlConverter/PhantomMagickConverter.php]: missing return type

File: src/Services/HtmlConverter/PhantomMagickConverter.php [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        $this->pages = $this->pages->push($html);
+
+        return $this;
+    }
+
+    public function save(string $path, string $type = 'pdf')
```
Comment (tychovbh) [inline src/Services/HtmlConverter/PhantomMagickConverter.php]: missing doc

Comment (tychovbh) [reply src/Services/HtmlConverter/PhantomMagickConverter.php]: return boolean true/false

File: tests/TestCase.php [hunk 31]
```diff
@@ -115,18 +114,39 @@ public function setUp(): void
...
+                    'DocuSign' => [
+
                     ]
                 ]
             ]
         );
 
+        Config::set('mvc-html-converter', [
+            'default' => 'PhantomMagickConverter',
+            'providers' => [
```
Comment (tychovbh) [inline tests/TestCase.php]: remove this

File: tests/TestCase.php [hunk 32]
```diff
@@ -115,18 +114,39 @@ public function setUp(): void
...
+                'PhantomMagickConverter' => [
+
+                ]
+            ]
+        ]);
+
+        Config::set('mvc-address-lookup', [
+            'default' => 'PdokService',
+            'providers' => [
```
Comment (tychovbh) [inline tests/TestCase.php:138]: update this

File: tests/feature/PaymentTest.php [hunk 33]
```diff
@@ -33,7 +33,9 @@ public function itCanStore()
         $data['products'] = $products;
         $payment->products = $products;
         $payment->status = Payment::STATUS_OPEN;
-        $response = $this->store('payments.store', PaymentResource::make($payment), $data);
+        $response = $this->store('payments.store', PaymentResource::make($payment), $data, 201, [
```
Comment (tychovbh) [inline tests/feature/PaymentTest.php]: revert this to original

File: tests/feature/services/PdokTest.php [hunk 34]
```diff
@@ -5,11 +5,20 @@
...
 
 use GuzzleHttp\Exception\GuzzleException;
+use Tychovbh\Mvc\Services\AddressLookup\AddressLookupInterface;
 use Tychovbh\Mvc\Services\AddressLookup\PdokService;
 use Tychovbh\Tests\Mvc\TestCase;
 
 class PdokTest extends TestCase
 {
+
+    public function __construct($name = null, array $data = [], $dataName = '')
```
Comment (tychovbh) [inline tests/feature/services/PdokTest.php]: remove this?

File: tests/feature/services/SignRequestTest.php [hunk 35]
```diff
@@ -1,14 +1,15 @@
...
 
-
 namespace Tychovbh\Tests\Mvc\feature\services;
 
-
 use GuzzleHttp\Client;
 use Illuminate\Http\UploadedFile;
 use Illuminate\Support\Arr;
 use Illuminate\Support\Facades\Storage;
+use Tychovbh\Mvc\Services\DocumentSign\DocumentSignInterface;
```
Comment (tychovbh) [inline tests/feature/services/SignRequestTest.php]: is this being used?

File: tests/feature/services/SignRequestTest.php [hunk 36]
```diff
@@ -1,14 +1,15 @@
...
 use Tychovbh\Mvc\Services\DocumentSign\SignRequest;
+use Tychovbh\Mvc\Services\HtmlConverter\HtmlConverterInterface;
```
Comment (tychovbh) [inline tests/feature/services/SignRequestTest.php]: is this being used?

**Review by tychovbh — CHANGES_REQUESTED** [review]

File: src/Contract.php (hunk 9 above)
Comment (thvrijn) [reply src/Contract.php]: There is no mvc-contract.. Don't you mean mvc-document-sign?

File: src/Services/DocumentSign/SignRequest.php (hunk 19 above)
Comment (thvrijn) [reply src/Services/DocumentSign/SignRequest.php]: What type should webhook be?

File: config/mvc-address-lookup.php [hunk 37]
```diff
@@ -4,7 +4,7 @@
     'default' => env('ADDRESS_LOOKUP_SERICE', 'PdokService'),
     'providers' => [
         'PdokService' => [
-
+            'base-url' => ''
```
Comment (tychovbh) [inline config/mvc-address-lookup.php]: always use underscores in configs and this should be filled by .env

File: config/mvc-document-sign.php [hunk 38]
```diff
@@ -4,8 +4,8 @@
     'default' => env('DOCUMENT_SIGN_PROVIDER', 'SignRequest'),
     'providers' => [
         'SignRequest' => [
-            'token' => '',
-            'subdomain' => '',
+            'token' => env('DOCUMENT_SIGN_PROVIDER_TOKEN', ''),
```
Comment (tychovbh) [inline config/mvc-document-sign.php]: this is not DOCUMENT_SIGN_PROVIDER but SIGNREQUEST_TOKEN

File: config/mvc-document-sign.php [hunk 39]
```diff
@@ -4,8 +4,8 @@
...
+            'subdomain' => env('DOCUMENT_SIGN_PROVIDER_SUBDOMAIN', ''),
```
Comment (tychovbh) [inline config/mvc-document-sign.php]: change

File: src/Contract.php (hunk 9 above)
Comment (tychovbh) [reply src/Contract.php]: no i mean mvc-contracts if it not exist create it

File: src/Services/AddressLookup/PdokService.php [hunk 40]
```diff
@@ -19,7 +19,7 @@ public static function search(string $zipcode, int $house_number): array
         $zipcode = str_replace(' ', '', $zipcode);
 
         $client = new Client();
-        $res = $client->request('GET', 'https://geodata.nationaalgeoregister.nl/locatieserver/v3/free', [
+        $res = $client->request('GET', config('mvc-address-lookup.providers.PdokService.base-url'), [
```
Comment (tychovbh) [inline src/Services/AddressLookup/PdokService.php]: set mvc-address-lookup.providers.PdokService as config property in constructer

File: src/Services/DocumentSign/SignRequest.php [hunk 41]
```diff
@@ -21,6 +21,11 @@ class SignRequest implements DocumentSignInterface
      */
     private $signers;
 
+    /**
+     * @var Arr
```
Comment (tychovbh) [inline src/Services/DocumentSign/SignRequest.php]: Arr should be Array

File: src/Services/DocumentSign/SignRequest.php [hunk 42]
```diff
@@ -63,17 +75,17 @@ private function createRequest(string $file, $name, $webhook = null)
...
+            error('SignRequestService create error', [
                 'message' => $exception->getMessage(),
                 'line' => $exception->getLine(),
             ]);
         }
     }
 
     /**
-     * Creates a Document
+     * Creates a Documentq
```
Comment (tychovbh) [inline src/Services/DocumentSign/SignRequest.php]: documentq ? q?

File: src/Services/HtmlConverter/PhantomMagickConverter.php [hunk 43]
```diff
@@ -12,23 +12,31 @@ class PhantomMagickConverter implements HtmlConverterInterface
...
-    public function page(string $html)
+    public function page(string $html): HtmlConverterInterface
     {
         $this->pages = $this->pages->push($html);
 
         return $this;
     }
 
-    public function save(string $path, string $type = 'pdf')
+    /**
```
Comment (tychovbh) [inline src/Services/HtmlConverter/PhantomMagickConverter.php:35]: missing description check whole pull request on doc blocks and description

**Review by tychovbh — CHANGES_REQUESTED** [review]

File: config/mvc-address-lookup.php (hunk 37 above)
Comment (tychovbh) [reply config/mvc-address-lookup.php]: and there is an spelling error in ADDRESS_LOOKUP_SERICE

**Pull request merged**